pkg/aws/eni/routing/info.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package enirouting
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/cilium/cilium/pkg/mac"
+)
+
+// RoutingInfo represents information that's required to enable
+// connectivity via the local rule and route tables while in ENI mode. The
+// information in this struct is used to create rules and routes which direct
+// traffic out of the ENI devices (egress).
+//
+// This struct is mostly derived from the `ipam.AllocationResult` as the
+// information comes from IPAM.
+type RoutingInfo struct {
+	// IPv4Gateway is the gateway where outbound/egress traffic is directed.
+	IPv4Gateway net.IP
+
+	// IPv4CIDRs is a list of CIDRs which the ENI device has access to. In most
+	// cases, it'll at least contain the CIDR of the IPv4Gateway IP address.
+	IPv4CIDRs []net.IPNet
+
+	// MasterIfMAC is the MAC address of the master interface that egress
+	// traffic is directed to. This is the MAC of the ENI itself which
+	// corresponds to the IPv4Gateway IP addr.
+	MasterIfMAC mac.MAC
+}
+
+// NewRoutingInfo creates a new RoutingInfo struct from a source which adheres
+// to the getter interface. The data from the source is then parsed and
+// validated.
+func NewRoutingInfo(getter getter) (*RoutingInfo, error) {
```

Comment:
@joestringer I've tried out your suggestion from: https://github.com/cilium/cilium/pull/11073#discussion_r411797976

After implementing it, I'm not sure what benefits we get from abstracting away the source of the info via the `getter` interface. I think we'd get the same functionality and non-code-interdependence from just having the `NewRoutingInfo` take the raw values as parameters. 

Is there something I'm missing that the interface would buy us?

Reply:
Thanks for trying it out, good question :)

Looking closer, in this case the ability to fetch the routing info, ie the implementation of `getter` is trivial; it's just taking the API response structure and returning the values inside. So in this current implementation I'm inclined to agree that the benefit is not obvious. Right now, it allows `pkg/aws/eni/routing` to avoid importing the API structures, but I don't think that's a big deal; we're already assuming a model around the cloud IPAM API responses in all of the ENI code structure anyway so the abstraction doesn't help.

If the full implementation of that getter was more complex (For example actually involved pulling values from an API) then I think it would make a bigger difference.

Looks like we got more value out of the `routingConfigurer` above, which allowed us to remove ENI references from that package (module the variable names, which should probably be simplified).

Reply:
Oh, perhaps the unsaid piece above: With this style of implementation, you can create your own dummy `getter` and actually unit-test the `parse()` function below trivially. In this specific case, if we remove it, then you just have to instantiate IPAM API response structures in the tests which is probably still straightforward.

If the `getter` was more complex then just to unit test this functionality it would be more work.

Reply:
Right. The retrieval of these ENI routing values is trivial and will probably continue to be, since it's inherently a cloud API interaction. So our `getter` isn't helping by providing any useful abstraction.

I'm going to take out the `getter` and replace the constructor with raw string values as input instead. This will still allow us to unit-test `parse` easily since it will just take those values as well.

Reply:
> (module the variable names, which should probably be simplified).

Which variable names need to be simplified?

Reply:
Oh, I mean `eniHealthRouting` -> `routeConfiguration`. This assumes an eni implementation but it doesn't make sense to assume that when a generic routeConfigurer is handed down.